src/compiler/crystal/command/playground.cr

```diff
@@ -34,6 +34,21 @@ class Crystal::Command
       end
     end
 
+    playground_url = "http://#{server.host || "localhost"}:#{server.port || "8080"}"
```

Comment:
`8080` is already the default (https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/tools/playground/server.cr#L460)

Reply:
I even think  `"localhost"` can to be dropped to when looking at https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/tools/playground/server.cr#L549 .

You can change to `property host : String = Socket::IPAddress::LOOPBACK` in https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/tools/playground/server.cr#L453

Reply:
Thanks for the review @j8r!

To be clear: 

1. Can I just remove the ` || "8080"` from `server.port || "8080"`?

2. Can I just remove ` || "localhost"` from `server.host || "localhost"`? I'll have to check later tonight but as I recall the value of `server.host` was not set.

Thanks and sorry for my newbness.

Reply:
You delete both yes. `server.host` is nillable (`String?`), with `nil` as default value. You have to change it to `String` in `Playground::Server`, with `Socket::IPAddress::LOOPBACK` as default value. See my links above to have hints where changing the code.

Reply:
Ah, gotcha. I'll make those changes and get you to look at it. Thanks again.

Reply:
The exact address to connect to is only known after the socket is bound. For example, with `--port 0`, the port is chosen by the socket implementation.

So this needs to be moved after `server.start`.

Reply:
@straight-shoota If I move the `open_browser` call after the `server.start` it never gets called, or at least on macOS the browser window never launches. Any way on getting around that?

Reply:
@j8r When I change `property host : String?` to `property host : String = Socket::IPAddress::LOOPBACK` I get the following errors when running `make`: 

```bash
Error in src/compiler/crystal.cr:8: instantiating 'Crystal::Command.class#run()'

Crystal::Command.run
                 ^~~

in src/compiler/crystal/command.cr:45: instantiating 'run(Array(String))'

  def self.run(options = ARGV)
  ^

in src/compiler/crystal/command.cr:46: instantiating 'Crystal::Command#run()'

    new(options).run
                 ^~~

in src/compiler/crystal/command.cr:70: instantiating 'playground()'

      playground
      ^~~~~~~~~~

in src/compiler/crystal/command/playground.cr:8: instantiating 'Crystal::Playground::Server.class#new()'

    server = Playground::Server.new
                                ^~~

in src/compiler/crystal/tools/playground/server.cr:459: instance variable '@host' of Crystal::Playground::Server must be String, not Nil

      @host = nil
      ^~~~~

make: *** [.build/crystal] Error 1
```

Do I need to set `@host = nil` to `@host = Socket::IPAddress::LOOPBACK`?

If so, something very strange happens and I'm unsure if it's related or just a coincidence.

First off, the `make` command completes, so that's great, but then when I run `bin/crystal play` the crystal text that is in the `playground` is different than before the change. See the following comparison (left is before change and right is after change):

![Screen Shot 2019-03-13 at 6 00 46 PM](https://user-images.githubusercontent.com/180819/54322224-fac52080-45b9-11e9-806c-ecc7444d4554.png)

Anyway, let me know what your thoughts are on this before I proceed with removing the fallback to `localhost`. 

Thanks.

Reply:
Remove the `@host = nil`

Reply:
You must not guess the adress. Use the one the server is bound to. `server.start` not only binds the socket but also listens wich will block the program. Execute launch code between bind and listen, for example by yielding to a block.

Reply:
To set the host and the port:
```cr
if host
  server.host = host
end
if port
  server.port = port
end
```

You can yield the playground address, then inside the block open the browser. In `Playground::Server` the yield can be just before the `server.listen`.